File: web/packages/alloy-storage/src/core/sign-in-result.ts
/** Why a sign-in failed — the phase, not user-facing copy. */
export type SignInFailureReason =
  | 'configurationInvalid' // bad auth URL / missing or corrupt pending-session entry / no auth UI
  | 'stateMismatch' // CSRF state check failed, or no code in the callback
  | 'exchangeFailed' // token endpoint rejected or unreachable
  | 'vaultFailed'; // token persistence failed (IndexedDB / Keychain)

/** Result of completeSignIn (web) / signIn (Apple). `cancelled` is a normal
 *  outcome, not an error; `detail` is for developers (logs, status lines),
 *  never end-user copy. */
export type SignInResult =
  | { outcome: 'success' }
  | { outcome: 'cancelled' }
  | { outcome: 'failed'; reason: SignInFailureReason; detail: string; status?: number };

File: swift/Sources/AlloyStorage/Core/SignInResult.swift
/// Why a sign-in failed — the phase, not user-facing copy.
public enum SignInFailureReason: String, Sendable {
  case configurationInvalid // bad auth URL / no auth UI available on this platform
  case stateMismatch // CSRF state check failed, or no code in the callback
  case exchangeFailed // token endpoint rejected or unreachable
  case vaultFailed // token persistence failed (Keychain)
}

/// Result of signIn (Apple) / completeSignIn (web). `cancelled` is a normal
/// outcome, not an error; `detail` is for developers (logs, status lines),
/// never end-user copy. Twin of core/sign-in-result.ts.
public enum SignInResult: Equatable, Sendable {
  case success
  case cancelled
  case failed(reason: SignInFailureReason, detail: String, status: Int?)
}
